File: ios/TPPSDK.m
// TPPSDKModuleExposure.m

#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(TPPSDK, RCTEventEmitter)

// Expose the module to React Native
// The first parameter is the name of the Swift class, and the second is the name you want to use in JavaScript.
// If the second parameter is nil, it uses the Swift class name by default.


// Expose methods
// Replace "methodName" with your actual method names
RCT_EXTERN_METHOD(startDeviceDiscovery:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(cancelDeviceDiscovery)
RCT_EXTERN_METHOD(connect:(NSString *)deviceName timeout:(nonnull NSNumber *)timeout callback:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(disconnect)
RCT_EXTERN_METHOD(startTransaction:(NSString *)amount callback:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(getSerialNumber:(RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(cancelTransaction)
RCT_EXTERN_METHOD(initialize:(NSString *)apiKey)

@end

File: ios/TPPSDK.swift
// MagTekModule.swift

import Foundation
import MagTekSDK

@objc(TPPSDK)
class TPPSDK:  RCTEventEmitter {
  private static var reader: MagTekCardReader?
  private var transactionCallbackCalled: Bool = false

  @objc func initialize(_ apiKey: String) {
    TPPSDK.reader = MagTekCardReader(apiKey, debug: true)
  }

  @objc override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  override func supportedEvents() -> [String]! {
    return ["TransactionUpdate"]
  }

  @objc func startDeviceDiscovery(_ callback: @escaping RCTResponseSenderBlock) {
    TPPSDK.reader?.startDeviceDiscovery { name, rssi in
      callback([["name": name, "rssi": rssi]])
    }
  }

  @objc func cancelDeviceDiscovery() {
    TPPSDK.reader?.cancelDeviceDiscovery()
  }

  @objc func connect(_ deviceName: String, timeout: TimeInterval, callback: @escaping RCTResponseSenderBlock) {
    TPPSDK.reader?.connect(deviceName) { connected in
      print("connected: \(connected)")
      callback([connected])
    }
  }

  @objc func disconnect() {
    TPPSDK.reader?.disconnect()
  }

  @objc func startTransaction(_ amount: String) {
    TPPSDK.reader?.startTransaction(amount) { [weak self] message, event, status in
      let transactionInfo = [
        "message": message,
        "event": MagTekCardReader.getEventMessage(event),
        "status": MagTekCardReader.getStatusMessage(status)
      ]
      self?.sendEvent(withName: "TransactionUpdate", body: transactionInfo)
    }
  }

  @objc func getSerialNumber(_ callback: RCTResponseSenderBlock) {
    let serialNumber = TPPSDK.reader?.getSerialNumber() ?? ""
    callback([serialNumber])
  }

  @objc func cancelTransaction() {
    TPPSDK.reader?.cancelTransaction()
  }
}
